packages/ui/VirtualList/VirtualListBase.js

```diff
@@ -23,728 +27,771 @@ const gridListItemSizeShape = PropTypes.shape({
-				overhang !== nextProps.overhang ||
-				spacing !== nextProps.spacing
-			);
 
-		this.hasDataSizeChanged = (dataSize !== nextProps.dataSize);
+		constructor (props) {
+			super(props);
 
-		if (hasMetricsChanged) {
-			this.calculateMetrics(nextProps);
-			this.updateStatesAndBounds(nextProps);
-		} else if (this.hasDataSizeChanged) {
-			this.updateStatesAndBounds(nextProps);
+			this.state = {firstIndex: 0, numOfItems: 0};
 		}
-	}
-
-	scrollBounds = {
-		clientWidth: 0,
-		clientHeight: 0,
-		scrollWidth: 0,
-		scrollHeight: 0,
-		maxLeft: 0,
-		maxTop: 0
-	}
 
-	moreInfo = {
-		firstVisibleIndex: null,
-		lastVisibleIndex: null
-	}
+		componentWillMount () {
+			if (this.props.clientSize) {
+				this.calculateMetrics(this.props);
+				this.updateStatesAndBounds(this.props);
+			}
+		}
 
-	primary = null
-	secondary = null
+		// Calculate metrics for VirtualList after the 1st render to know client W/H.
+		// We separate code related with data due to re use it when data changed.
+		componentDidMount () {
+			if (!this.props.clientSize) {
+				this.calculateMetrics(this.props);
+				this.updateStatesAndBounds(this.props);
+			}
+			this.setContainerSize();
+		}
 
-	isPrimaryDirectionVertical = true
-	isItemSized = false
+		// Call updateStatesAndBounds here when dataSize has been changed to update nomOfItems state.
+		// Calling setState within componentWillReceivePropswill not trigger an additional render.
+		componentWillReceiveProps (nextProps) {
+			const
+				{dataSize, direction, itemSize, overhang, spacing} = this.props,
+				hasMetricsChanged = (
+					direction !== nextProps.direction ||
+					((itemSize instanceof Object) ? (itemSize.minWidth !== nextProps.itemSize.minWidth || itemSize.minHeight !== nextProps.itemSize.minHeight) : itemSize !== nextProps.itemSize) ||
+					overhang !== nextProps.overhang ||
+					spacing !== nextProps.spacing
+				);
+
+			this.hasDataSizeChanged = (dataSize !== nextProps.dataSize);
+
+			if (hasMetricsChanged) {
+				this.calculateMetrics(nextProps);
+				this.updateStatesAndBounds(nextProps);
+				this.setContainerSize();
+			} else if (this.hasDataSizeChanged) {
+				this.updateStatesAndBounds(nextProps);
+				this.setContainerSize();
+			}
+		}
 
-	dimensionToExtent = 0
-	threshold = 0
-	maxFirstIndex = 0
-	curDataSize = 0
-	hasDataSizeChanged = false
-	cc = []
-	scrollPosition = 0
-	updateFrom = null
-	updateTo = null
+		scrollBounds = {
+			clientWidth: 0,
+			clientHeight: 0,
+			scrollWidth: 0,
+			scrollHeight: 0,
+			maxLeft: 0,
+			maxTop: 0
+		}
 
-	containerRef = null
+		moreInfo = {
+			firstVisibleIndex: null,
+			lastVisibleIndex: null
+		}
 
-	isVertical = () => this.isPrimaryDirectionVertical
+		primary = null
+		secondary = null
 
-	isHorizontal = () => !this.isPrimaryDirectionVertical
+		isPrimaryDirectionVertical = true
+		isItemSized = false
 
-	getScrollBounds = () => this.scrollBounds
+		dimensionToExtent = 0
+		threshold = 0
+		maxFirstIndex = 0
+		lastFirstIndex = 0
```

Comment:
`lastFirstIndex` could be confused. How `prevFirstIndex`?

Reply:
Good! Updated.